spec.html

```diff
@@ -23,31 +23,64 @@
-      1. Let _replacer_ be ? GetMethod(_searchValue_, @@replace).
-      1. If _replacer_ is not *undefined*, then
-        1. Return ? Call(_replacer_, _searchValue_, &laquo; _O_, _replaceValue_ &raquo;).
-    1. Let _string_ be ? ToString(_O_).
-    1. Let _searchString_ be ? ToString(_searchValue_).
-    1. Let _functionalReplace_ be IsCallable(_replaceValue_).
-    1. If _functionalReplace_ is *false*, then
-      1. Set _replaceValue_ to ? ToString(_replaceValue_).
-    1. Search _string_ for the first occurrence of _searchString_ and let _pos_ be the index within _string_ of the first code unit of the matched substring and let _matched_ be _searchString_. If no occurrences of _searchString_ were found, return _string_.
-    1. If _functionalReplace_ is *true*, then
-      1. Let _replValue_ be ? Call(_replaceValue_, *undefined*, &laquo; _matched_, _pos_, _string_ &raquo;).
-      1. Let _replStr_ be ? ToString(_replValue_).
-    1. Else,
-      1. Let _captures_ be a new empty List.
-      1. Let _replStr_ be GetSubstitution(_matched_, _string_, _pos_, _captures_, *undefined*, _replaceValue_).
-    1. Let _tailPos_ be _pos_ + the number of code units in _matched_.
-    1. Let _newString_ be the string-concatenation of the first _pos_ code units of _string_, _replStr_, and the trailing substring of _string_ starting at index _tailPos_. If _pos_ is 0, the first element of the concatenation will be the empty String.
-    1. Return _newString_.
-  </emu-alg>
-  <emu-note>
-    <p>The `replaceAll` function is intentionally generic; it does not require that its *this* value be a String object. Therefore, it can be transferred to other kinds of objects for use as a method.</p>
-  </emu-note>
+<pre>
+StringIndexOf(string, searchValue, fromIndex)
+Note: OOB fromIndex values (fromIndex >= string.length) are accepted by the algorithm.
+
+1. Assert: Type(string) is String.
+2. Assert: Type(searchValue) is String.
+3. Let length be the number of code units in string.
+4. Assert: fromIndex is a nonnegative integer.
+5. Search string for the first occurrence of searchValue s.t. the index within string of the first code unit of the matched substring is >= fromIndex. Let pos be that index if such an occurrence is found; otherwise let pos be -1.
+6. Return pos.
+
+String.p.replaceAll(searchValue, replaceValue)[a]
+
+1. Let O be ? RequireObjectCoercible(this value).
+2. If searchValue is neither undefined nor null, then
+   1. Let replacer be ? GetMethod(searchValue, @@replaceAll).
+   2. If replacer is not undefined, then
+      1. Return ? Call(replacer, searchValue, « O, replaceValue »).
+1. Let string be ? ToString(O).
+2. Let searchString be ? ToString(searchValue).
+3. Let functionalReplace be IsCallable(replaceValue).
+4. If functionalReplace is false, then
+   1. Let replaceValue be ? ToString(replaceValue).
+1. // Collect all matches:
+2. Let searchLength be the length of searchString.
+3. Let advanceBy be max(1, searchLength).[b][c]
```

Comment:
shouldn't this use https://tc39.github.io/ecma262/#sec-advancestringindex?

Reply:
I asked myself that as well, according to @mathiasbynens we must not consider surrogates here.

Reply:
If the forged has a u flag, we would; why wouldn’t we want the string to behave as if it has a u flag?

Reply:
This would be a deviation from `replace`, which the committee previously stated we must match as closely as possible.

```js
const lead = '\uD834';
const trail = '\uDF06';
const pair = `${lead}${trail}`;
console.assert('𝌆' === '\uD834\uDF06');
console.assert(pair.replace(lead, 'x') === `x${trail}`);
```

The current proposal matches `replace` in this regard:

```js
console.assert(pair.replaceAll(lead, 'x') === `x${trail}`);
```

Changing it would mean surrogate halves are only matched if they appear as lone surrogates in the input string:

```js
console.assert(pair.replaceAll(lead, 'x') === pair);
```